src/components/ingestion/AbstractIngestionEngine.js

```diff
@@ -1,100 +1,209 @@
-	 * @return {ReadableStream}                The input stream (for chaining)
 	 */
-	recordStream = (readableStream) => readableStream
-		.on('data', async (data) => {
-			readableStream.pause()
-			await this.ingestData(data)
-			readableStream.resume()
+	processMpegtsStreamData = (mpegtsData, enc, done) => {
+		this.mpegtsDemuxer.process(mpegtsData)
+		const demuxedPacket = this.getMostRecentDemuxedPacket() || generateEmptyPacket()
+		const payload = new Payload({
+			data: mpegtsData,
+			type: dataTypes.STREAM.CONTAINER,
+			duration: 0,
+			position: tsToMilliseconds(demuxedPacket.pts),
+			createdAt: Date.now(),
 		})
+		done(null, payload)
+	}
 
 	/**
 	 * Start the ingestion engine.
 	 *
-	 * @return {null}
+	 * This creates an FFmpeg process configured to convert input into a predetermined
+	 * format as specified by getFfmpegSettings, and announce new data to the ingestion
+	 * engine.
+	 *
+	 * It then spawns the input stream as specified by the implementing class, and creates
+	 * a pipeline that passes input->FFmpeg->packaging->ingestion
+	 *
+	 * @return {Stream} the ingestion pipeline that just got started
 	 */
 	start = () => {
-		const ffmpegSettings = this.getFfmpegSettings()
-		const ffmpegProcess = spawn('ffmpeg', ffmpegSettings)
+		this.ffmpegProcess = spawn('ffmpeg', this.getFfmpegSettings())
 		const inputStream = this.getInputStream()
-		if (inputStream !== null) {
-			inputStream.pipe(ffmpegProcess.stdin)
-		}
-		this.recordStream(ffmpegProcess.stdout)
+
 		logger.info(`Starting ingestion from ${this.constructor.name}...`)
+		inputStream.pipe(this.ffmpegProcess.stdin)
+		this.activeIngestionPipeline = pipeline(
+			this.ffmpegProcess.stdout,
+			this.mpegtsProcessingStream,
+			this.payloadIngestionStream,
+			() => this.stop(),
+		)
+		return this.activeIngestionPipeline
 	}
 
 	/**
-	 * Returns an ffmpeg settings array for this ingestion engine.
+	 * Stop the ingestion engine.
 	 *
-	 * @return {String[]} A list of ffmpeg command line parameters.
+	 * This stops the ingestion stream if it exists and terminates the FFmpeg process if it exists
 	 */
-	getFfmpegSettings = () => [
-		'-i', this.getInputLocation(),
-		...this.getTranscodeSettings(),
-		'-f', 'mpegts',
-		'-',
-	]
+	stop = () => {
+		if (this.activeIngestionPipeline !== null) {
+			this.activeIngestionPipeline.end()
+		}
+		if (this.ffmpegProcess !== null) {
+			this.ffmpegProcess.kill()
+		}
+		logger.info(`Ended ingestion from ${this.constructor.name}...`)
+	}
 
 	/**
-	 * Returns the input to be passed to ffmpeg's `-i` parameter.
-	 *
-	 * NOTE: If this is not overridden then getInputStream must be implemented.
+	 * Returns an FFmpeg settings array for this ingestion engine.
 	 *
-	 * @return {String} The input location to be passed to ffmpeg
+	 * @return {String[]} A list of FFmpeg command line parameters.
 	 */
-	getInputLocation = () => '-'
+	getFfmpegSettings = () => [
+		'-loglevel', 'info',
+		'-i', '-',
+		'-f', 'mpegts',
```

Comment:
It's nominally outside of the scope of this PR, but it may be extremely worthwhile to go ahead and add the [Stream Copy settings](https://ffmpeg.org/ffmpeg.html#Stream-copy) (`-codec copy`), since currently we are silently changing the encoding of the stream, and that is not intended.

(See earlier comment; default FFmpeg behavior without the stream copy command is to re-encode into default formats, which are something like H264 for video and no idea for audio.)

Reply:
This PR has so much out of scope (e.g. a complete rewrite of the way streams are processed!) that I think this should be fine!